### DISCUSSION TITLE
Passing config for Intelephense PHP

### DISCUSSION THREAD

**Discussion by Kevinizevbigie** [joaotavora/eglot#1365] [Q&A] [answered]
I am not quite sure this is a bug, so I opened a discussion instead. 

I use eglot with php `intelephense`. I have tried various combinations to pass configuration for intelephense via `eglot-workspace-configuration`. It does not work.

## Reproduction recipe

1. Ensure the intelephense server is installed and available in path
2. Add eglot intelephense to your `init.el`
```lisp
(use-package eglot
  :ensure t
  :config
  (with-eval-after-load 'eglot
    (setq completion-category-defaults nil)
    (add-to-list 'eglot-server-programs
                     '((php-mode web-mode)
                       . ("intelephense" "--stdio" :initializationOptions
                          (:globalStoragePath "/home/kev/.tmp"))))
```
This works perfectly.

3. Add the workspace configuration to switch `intelephense` OFF.
```lisp
(setq-default eglot-workspace-configuration
              '((:intelephense . ((diagnostics.enable nil)))))
```
4. Open a php file and run eglot. Wait for LSP to start.

## The issue
The lsp is still reporting diagnostic errors despite passing the config to switch it off.

## Combinations

I have tried various combinations for eglot-workspace-configuration, for example:

```lisp
(setq-default eglot-workspace-configuration
              '((:intelephense . (:diagnostics . (enable nil)))))
```

*How do I do this properly for PHP/Intelephense?*

The json verson of the config would look like this:

```json
{
  "intelephense.diagnostics.enable": true,
}
```

**Comment by joaotavora** [comment]
Go into the Eglot events buffer, which in very recent versions reports communication in its final JSON format, and find the place where Eglot sends over this info to the server. Then paste it here. Refer to the manual for how to reach that buffer.

You may have too many dots in your Elisp, it's supposed to be a plist.

**Reply by Kevinizevbigie** [reply, marked as answer]
Nice. The event buffer gives some very valuable insights. I fixed it. 

There were two things wrong:

1. I am a lisp programmer mainly and so I default to `nil` for false. Reading over the docs again I saw that `:json-false` translates to `false`. And `nil` translated to `null` in json. Question, why not stick to `t` and `nil`?
2. Secondly, the correct workspace block is:

```lisp
(setq-default eglot-workspace-configuration
              '(:intelephense (:diagnostics (:enable :json-false))))
```

Thanks @joaotavora , great lsp.

**Reply by joaotavora** [reply]
> Question, why not stick to t and nil?

Because nil is conflated to the empty list in Lisp.  If you used it for JSON `false`, you'd need some other lisp form for JSON `null` and empty JSON object anyway.  The correspondence chosen was one among many, chosen early on.  Perhaps not the best, but not the worst either. Vhanging it would break lots and lots of users.

**Reply by Kevinizevbigie** [reply]
Agreed. I think it makes clear sense once one knows that it is.
